File: 2025/9/9_23_leetcode_bc/102.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def levelOrder(self, root: Optional[TreeNode]) -> List[List[int]]:
        from collections import deque

        if not root:
            return []

        current_nodes = deque([root])
        next_nodes = deque()
        output = []
        level_vals = []

        while current_nodes:
            cur_node = current_nodes.popleft()
            level_vals.append(cur_node.val)

            if cur_node.left:
                next_nodes.append(cur_node.left)
            if cur_node.right:
                next_nodes.append(cur_node.right)

            if not current_nodes:
                output.append(level_vals)
                level_vals = []
                current_nodes, next_nodes = next_nodes, deque()

        return output

File: 2025/9/9_23_leetcode_bc/102.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @return {number[][]}
 */
var levelOrder = function(root) {
    if (!root) return [];

    let currentNodes = [root];
    let nextNodes = [];
    let output = [];
    let level_vals = [];

    while (currentNodes.length > 0 ) {
        let node = currentNodes.shift();
        level_vals.push(node.val);

        if (node.left) nextNodes.push(node.left);
        if (node.right) nextNodes.push(node.right);

        if (currentNodes.length === 0) {
            output.push(level_vals);
            level_vals = [];
            currentNodes = nextNodes;
            nextNodes = [];
        }
    }

    return output;
};